sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,207 @@
+                }
+                else if (OperationFailedException != null)
+                {
+                    throw OperationFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(Value));
+                }
+
+                _value = value;
+                HasValue = true;
+            }
+        }
+
+        public Response RawResponse { get; set; }
+
+        public string OperationTypeName { get; set; }
+
+        public TimeSpan DefaultPollingInterval { get; set; }
+
+        private ClientDiagnostics ClientDiagnostics { get; }
+
+        private RequestFailedException OperationFailedException
+        {
+            get => _operationFailedException;
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                _operationFailedException = value;
+            }
+        }
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken) =>
+            await WaitForCompletionAsync(DefaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                var serverDelay = response.Headers.TryGetValue(RetryAfterHeaderName, out string retryAfterValue)
+                    && int.TryParse(retryAfterValue, out int serverDelayInSeconds)
+                    ? TimeSpan.FromSeconds(serverDelayInSeconds) : TimeSpan.Zero;
+
+                var delay = serverDelay > pollingInterval
+                    ? serverDelay : pollingInterval;
+
+                await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
```

Comment:
Lets cache this value.

Reply:
Are you talking about `$"{OperationTypeName}.UpdateStatus"`? We can do it.

By the way, what do you think about moving `OperationTypeName` to the constructor instead of making it settable? Ideally, we expect developers to set it at most once. I made it as a property since it won't be used much. We could make it an optional parameter, though.

Reply:
>Are you talking about $"{OperationTypeName}.UpdateStatus"? We can do it.

Yeah, so we don't allocate per method call.

>By the way, what do you think about moving OperationTypeName to the constructor instead of making it settable?

I think it's a great idea. 👍